feat: better inferred typings

PalmDevs opened the pull request:
Makes patching functions a little more tolerable.

*(excuse my nonsensical examples)*
![Preview of change](https://github.com/user-attachments/assets/72c9c61c-e5aa-4566-adc0-288a36b08ed7)
![Preview of change](https://github.com/user-attachments/assets/1c242a03-de13-444e-97c3-6b56c8d93adc)

PalmDevs committed:
feat: better inferred typings

twnlink commented on src/shared.ts:
```diff
@@ -1,28 +1,42 @@
   p: WeakRef<any>;
   // cleanups
-  c: Function[];
+  c: AnyFunction[];
```
This is an array of cleanup functions. There is no practical benefit to using the `AnyFunction` type here over just `Function`.

twnlink commented on src/shared.ts:
```diff
@@ -1,28 +1,42 @@
 
+export type PatchTypeToCallbackMap<F extends AnyFunction> = {
+  a: (args: Parameters<F>, ret: ReturnType<F>) => ReturnType<F>;
+  b: (args: Parameters<F>) => ReturnType<F>;
```
Before patches may return an array to replace the function's original arguments. They cannot replace the function's return value, so this type definition is explicitly wrong.

twnlink commented:
I've given a few comments in my review for you to resolve.
Unrelated, but noting that your fork lives on the `revenge-mod` organization, I would highly recommend making a hard fork of spitroast for your use-case. Discord's React Native configuration is not setup to have WeakRef, which spitroast currently relies on. You also cannot polyfill WeakRef.

My friend is currently maintaining a fork of spitroast for these reasons, I have added modifications to it to account for usage within Hermes.

These two commits have the relevant changes to make spitroast work on mobile:
https://github.com/marshift/strawberry/commit/8af37a921f0d5f92a46e2f8ff1ee36de5fb11b01
https://github.com/marshift/strawberry/commit/33ec62f3d827b759c1d43c3ae26510e7cd66a92e

Please note that holding an `unpatch` function *will* hold the patched function in memory, this is unavoidable without WeakRef. Blame Hermes, not me.

PalmDevs committed:
fix: apply review changes

- Correct type for `before` hooks
- Change cleanup functions type to `Function[]`
